src/gsSolver/gsPreconditioner.h

```diff
@@ -0,0 +1,238 @@
+/** @file gsPreconditioner.h
+
+    @brief Simple abstract class for (discrete) linear preconditioners.
+
+    This file is part of the G+Smo library.
+
+    This Source Code Form is subject to the terms of the Mozilla Public
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+    Author(s): J. Sogn, A. Manzaflaris, C. Hofreither, S. Takacs, C. Hofer
+*/
+#pragma once
+
+#include <gsCore/gsLinearAlgebra.h>
+#include <gsIO/gsOptionList.h>
+#include <gsSolver/gsLinearOperator.h>
+
+namespace gismo
+{
+
+/// @brief Simple abstract class for perconditioners.
+///
+/// The class represents an iteration method in the form
+///
+/// \f$ x_{new} = x_{old} + P(f - A*x_{old}).\f$
+///
+/// If the number of steps is set to 1, the member function \a apply represents
+/// the application of \f$ P \f$. If \a setNumOfSweeps is used to set the
+/// number of sweeps to some \f$ \nu>0 \f$, the member function \a apply
+/// realizes \f$ I-(I-PA)^{\nu} A^{-1} \f$. The same interpretation applies
+/// if the object is used as a gsLinearOperator.
+///
+/// The member function \a step represents one step of the above iteration method.
+///
+/// Usually, the step operation can be performed in an optimized way.
+///
+/// The derived classes have to contain the functions: \a step, \a stepT,
+/// \a rows, \a cols.
+///
+/// \ingroup Solver
+template<class T>
+class gsPreconditionerOp : public gsLinearOperator<T>
+{
+public:
+
+    /// Shared pointer for gsLinearOperator
+    typedef memory::shared_ptr<gsPreconditionerOp> Ptr;
+
+    /// Unique pointer for gsLinearOperator
+    typedef memory::unique_ptr<gsPreconditionerOp> uPtr;
+
+    /// Base class
+    typedef gsLinearOperator<T> Base;
+
+    /// Base class
+    typedef typename gsLinearOperator<T>::Ptr BasePtr;
+
+    gsPreconditionerOp() : m_num_of_sweeps(1) {}
+
+    /**
+     * @brief Apply the method for given right hand side and current iterate
+     * @param rhs Right hand side vector
+     * @param x   Current iterate vector
+     */
+    virtual void step(const gsMatrix<T> & rhs, gsMatrix<T> & x) const = 0;
+
+    /**
+     * @brief Apply the transposed variant of the method for given right hand
+     *        side and current iterate
+     * @param rhs Right hand side vector
+     * @param x   Current iterate vector
+     */
+    virtual void stepT(const gsMatrix<T> & rhs, gsMatrix<T> & x) const { step( rhs, x); } // Assume symmetry.
```

Comment:
This symmetry assumption can be tricky/it might not hold. At lease some \warnings should be added in the doxygen docs of step() and stepT() and maybe in the class as well.